File: src/test/java/org/stt/persistence/BackupCreatorTest.java
package org.stt.persistence;

import org.apache.commons.io.FileUtils;
import org.apache.commons.io.filefilter.FileFileFilter;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import org.mockito.MockitoAnnotations;
import org.stt.config.BackupConfig;
import org.stt.config.ConfigRoot;
import org.stt.config.PathSetting;
import org.stt.time.DateTimes;

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.time.LocalDate;
import java.util.Collection;

import static org.hamcrest.CoreMatchers.is;

public class BackupCreatorTest {

	@Rule
	public TemporaryFolder tempFolder = new TemporaryFolder();

    private ConfigRoot configRoot = new ConfigRoot();
    private BackupConfig backupConfig = configRoot.getBackup();

    private File currentTempFolder;

	private File currentSttFile;
    private BackupCreator sut;

    @Before
    public void setup() throws IOException {
        MockitoAnnotations.initMocks(this);

		currentTempFolder = tempFolder.newFolder();
		currentSttFile = tempFolder.newFile();
        try (PrintWriter out = new PrintWriter(
                currentSttFile, "UTF8")) {
            out.print("blubb, just a test line");
        }

        backupConfig.setBackupRetentionCount(3);
        backupConfig.setBackupLocation(new PathSetting(currentTempFolder.getAbsolutePath()));

        sut = new BackupCreator(backupConfig, currentSttFile, "");
    }

	@Test
	public void existingBackupShouldPreventNewBackup() throws IOException {
		// GIVEN
        String threeDaysAgo = DateTimes.prettyPrintDate(LocalDate.now()
                .minusDays(3));
        String sttFileName = currentSttFile.getName();
        File backedUp = new File(currentTempFolder, sttFileName + "-"
				+ threeDaysAgo);
		createNewFile(backedUp);

		// WHEN
		sut.start();

		// THEN
		Collection<File> files = FileUtils.listFiles(currentTempFolder,
				FileFileFilter.FILE, null);

		Assert.assertEquals(1, files.size());
		Assert.assertThat(files.iterator().next().getAbsoluteFile(),
				is(backedUp.getAbsoluteFile()));
	}

	@Test
	public void oldBackupShouldBeDeleted() throws IOException {
		// GIVEN
        for (int i = 0; i < backupConfig.getBackupRetentionCount(); i++) {
            String xDaysAgo = DateTimes.prettyPrintDate(LocalDate.now()
                    .minusDays(i));
            String sttFileName = currentSttFile.getName();
            File oldFile = new File(currentTempFolder, sttFileName + "-"
					+ xDaysAgo);
			createNewFile(oldFile);
		}

        String xDaysAgo = DateTimes.prettyPrintDate(LocalDate.now()
                .minusDays(backupConfig.getBackupRetentionCount() + 1));
        String sttFileName = currentSttFile.getName();
        File oldFile = new File(currentTempFolder, sttFileName + "-" + xDaysAgo);
		createNewFile(oldFile);

		// WHEN
		sut.start();

		// THEN
		Assert.assertFalse("Old backup file should have been deleted",
				oldFile.exists());
	}

	@Test
	public void initialBackupShouldBeCreated() throws IOException {
		// GIVEN
        String currentDate = DateTimes.prettyPrintDate(LocalDate.now());
        String sttFileName = currentSttFile.getName();
        File expectedFile = new File(currentTempFolder, sttFileName + "-"
                + currentDate);

		// WHEN
		sut.start();

		// THEN
		Assert.assertTrue(
				"Original and backed up files do not have the same contents",
				FileUtils.contentEquals(currentSttFile, expectedFile));
	}

	@Test
	public void existingFileShouldNotBeOverwritten() throws IOException {
		// GIVEN
        String currentDate = DateTimes.prettyPrintDate(LocalDate.now());
        String sttFileName = currentSttFile.getName();
        File existingFile = new File(currentTempFolder, sttFileName + "-"
                + currentDate);
		createNewFile(existingFile);

		// WHEN
		sut.start();

		// THEN
		Assert.assertFalse(
				"Original and backed up files do not have the same contents",
				FileUtils.contentEquals(currentSttFile, existingFile));
	}

	private void createNewFile(File toCreate) throws IOException {
		Assert.assertTrue(
				"could not create test file " + toCreate.getAbsolutePath(),
				toCreate.createNewFile());
	}
}

File: src/test/kotlin/org/stt/persistence/BackupCreatorTest.kt
package org.stt.persistence

import org.apache.commons.io.FileUtils
import org.apache.commons.io.filefilter.FileFileFilter
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import org.mockito.MockitoAnnotations
import org.stt.config.ConfigRoot
import org.stt.config.PathSetting
import org.stt.time.DateTimes
import java.io.File
import java.io.PrintWriter
import java.nio.charset.StandardCharsets
import java.time.LocalDate

class BackupCreatorTest {

    @field:Rule
    @JvmField
    var tempFolder = TemporaryFolder()

    private val configRoot = ConfigRoot()
    private val backupConfig = configRoot.backup

    private var currentTempFolder: File? = null

    private lateinit var currentSttFile: File
    private var sut: BackupCreator? = null

    @Before
    fun setup() {
        MockitoAnnotations.initMocks(this)

        currentTempFolder = tempFolder.newFolder()
        currentSttFile = tempFolder.newFile()
        PrintWriter(currentSttFile, StandardCharsets.UTF_8.name()).use { out -> out.print("blubb, just a test line") }

        backupConfig.backupRetentionCount = 3
        backupConfig.backupLocation = PathSetting(currentTempFolder!!.absolutePath)

        sut = BackupCreator(backupConfig, currentSttFile, "")
    }

    @Test
    fun existingBackupShouldPreventNewBackup() {
        // GIVEN
        val threeDaysAgo = DateTimes.prettyPrintDate(LocalDate.now()
                .minusDays(3))
        val sttFileName = currentSttFile.name
        val backedUp = File(currentTempFolder, sttFileName + "-"
                + threeDaysAgo)
        createNewFile(backedUp)

        // WHEN
        sut!!.start()

        // THEN
        val files = FileUtils.listFiles(currentTempFolder!!,
                FileFileFilter.FILE, null)

        assertThat(1).isEqualTo(files.size.toLong())
        assertThat(files).first().isEqualTo(backedUp.absoluteFile)
    }

    @Test
    fun oldBackupShouldBeDeleted() {
        // GIVEN
        for (i in 0 until backupConfig.backupRetentionCount) {
            val xDaysAgo = DateTimes.prettyPrintDate(LocalDate.now()
                    .minusDays(i.toLong()))
            val sttFileName = currentSttFile.name
            val oldFile = File(currentTempFolder, sttFileName + "-"
                    + xDaysAgo)
            createNewFile(oldFile)
        }

        val xDaysAgo = DateTimes.prettyPrintDate(LocalDate.now()
                .minusDays((backupConfig.backupRetentionCount + 1).toLong()))
        val sttFileName = currentSttFile.name
        val oldFile = File(currentTempFolder, "$sttFileName-$xDaysAgo")
        createNewFile(oldFile)

        // WHEN
        sut!!.start()

        // THEN
        assertThat(oldFile).doesNotExist()
    }

    @Test
    fun initialBackupShouldBeCreated() {
        // GIVEN
        val currentDate = DateTimes.prettyPrintDate(LocalDate.now())
        val sttFileName = currentSttFile.name
        val expectedFile = File(currentTempFolder, sttFileName + "-"
                + currentDate)

        // WHEN
        sut!!.start()

        // THEN
        assertThat(currentSttFile).hasSameContentAs(expectedFile)
    }

    @Test
    fun existingFileShouldNotBeOverwritten() {
        // GIVEN
        val currentDate = DateTimes.prettyPrintDate(LocalDate.now())
        val sttFileName = currentSttFile.name
        val existingFile = File(currentTempFolder, sttFileName + "-"
                + currentDate)
        createNewFile(existingFile)

        // WHEN
        sut!!.start()

        // THEN
        assertThat(listOf(currentSttFile))
                .noneSatisfy { file ->
                    assertThat(file as File).hasSameContentAs(existingFile)
                }
    }

    private fun createNewFile(toCreate: File) {
        assertThat(toCreate.createNewFile()).isTrue()
    }
}
